File: vscode/src/protocol.ts
/**
 * HarnessGate 服务端的 WS 协议（与服务端 server/types.ts 对齐）。
 * 插件是瘦客户端：只发消息、收消息，agent 与文件都在服务器上。
 */

export type SessionStatus = "starting" | "ready" | "awaiting" | "saved" | "error" | "stopped";

export type ConfigOption = {
  id: string;
  name?: string;
  category?: string;
  currentValue?: string;
  options?: Array<{ value: string; name?: string; description?: string }>;
};

export type SessionInfo = {
  id: string;
  harnessId: string;
  harnessLabel: string;
  cwd: string;
  status: SessionStatus;
  createdAt: string;
  lastActiveAt: string;
  live: boolean;
  resumable: boolean;
  acpSessionId?: string;
  error?: string;
  /** 正在等用户点授权（此时 status === "awaiting"） */
  pendingPermission?: { requestId: string; title: string; options: PermissionOption[] };
  title?: string;
  modes?: { currentModeId?: string; availableModes?: Array<{ id: string; name?: string }> };
  configOptions?: ConfigOption[];
  worktree?: { dir: string; branch: string; baseCwd: string };
  /** 一个 turn 正在跑（面板据此显示「打断」按钮） */
  inTurn?: boolean;
  /** 权限自动决策档位（与网页版对齐；房间会话恒为 all 不可改） */
  autoApprove?: "off" | "readonly" | "all";
  roomId?: string;
};

export type HarnessAvailability = {
  id: string;
  label: string;
  available: boolean;
  binPath: string | null;
  note?: string;
  version?: string;
  description?: string;
  tier?: "vendor" | "known" | "unknown";
  trustReason?: string;
  blocked?: boolean;
  state?: string;
  configs?: ConfigOption[];
};

export type PermissionOption = { optionId: string; name: string; kind?: string };

export type TranscriptEntry =
  | { kind: "user"; text: string; ts: string }
  | { kind: "assistant"; text: string; ts: string }
  | { kind: "thought"; text: string; ts: string }
  | { kind: "tool"; title: string; status: string; ts: string; toolCallId?: string }
  | { kind: "permission"; title: string; ts: string; answered?: string; requestId?: string; options?: PermissionOption[]; auto?: boolean }
  | { kind: "error"; message: string; ts: string }
  | { kind: "log"; text: string; ts: string };

export type DirListing = {
  reqId: string;
  input: string;
  dir: string;
  exists: boolean;
  isDir: boolean;
  entries: Array<{ name: string; path: string; git: boolean }>;
  error?: string;
};

/* ---------- 圆桌（room）：与服务端 server/room.ts 对齐的子集 ---------- */

export type RoomTurn = {
  round: number;
  sessionId: string;
  harnessId: string;
  harnessLabel: string;
  prompt: string;
  reply: string;
  stopReason: string;
  ts: string;
  /** 主持人发言（开场/轮间小结/最终汇总）横跨整行 */
  kind?: "member" | "host" | "review";
  hostRole?: "opening" | "round-summary" | "final";
  crewTaskId?: string;
  score?: number;
  scoreNote?: string;
};

export type RoomTopic = {
  id: string;
  topic: string;
  rounds: number;
  mode: "parallel" | "sequential";
  converge?: boolean;
  tournament?: boolean;
  convergedRound?: number;
  status: "idle" | "running" | "done" | "error" | "stopped";
  turns: RoomTurn[];
  /** 正在进行的轮次（0 = 主持人开场阶段） */
  currentRound?: number;
  createdAt: string;
  error?: string;
};

export type RoomMember = { sessionId: string; harnessId: string; harnessLabel: string };

export type CrewTaskLite = {
  id: string;
  title: string;
  status: string;
  assignee?: string;
  files?: string[];
  summary?: string;
  review?: { reviewer: string; verdict: string; score?: number; comments: string };
};

export type Room = {
  id: string;
  topic: string;
  members: string[];
  memberInfo?: RoomMember[];
  host?: { sessionId: string; harnessId: string; harnessLabel?: string } & Record<string, unknown>;
  cwd?: string;
  rounds: number;
  mode: "parallel" | "sequential";
  converge?: boolean;
  tournament?: boolean;
  writeAllowed: boolean;
  status: "idle" | "running" | "done" | "error" | "stopped";
  turns: RoomTurn[];
  topics?: RoomTopic[];
  /** 工作队模式：任务板（简化视图，完整面板在网页版） */
  crew?: { phase: string; tasks: CrewTaskLite[]; mergeLines?: string[]; conflicts?: Array<{ branch: string }> } & Record<string, unknown>;
  createdAt: string;
  updatedAt: string;
  error?: string;
};

export type ClientMsg =
  | { type: "create"; harnessId: string; cwd?: string; isolate?: boolean }
  | { type: "resume"; sessionId: string }
  | { type: "mode"; sessionId: string; modeId: string }
  | { type: "set-auto-approve"; sessionId: string; level: "off" | "readonly" | "all" }
  | { type: "config"; sessionId: string; configId: string; value: string }
  | { type: "prompt"; sessionId: string; text: string }
  | { type: "permission"; sessionId: string; requestId: string; optionId: string }
  | { type: "close"; sessionId: string }
  /** 打断当前回合（session/cancel），会话保持可用；不同于 close（停整个会话进程） */
  | { type: "interrupt"; sessionId: string }
  | { type: "delete"; sessionId: string }
  | { type: "handoff"; sessionId: string; keep?: number }
  | { type: "sync-history"; harnessId?: string; force?: boolean }
  | { type: "transcript"; sessionId: string }
  | { type: "dirs"; reqId: string; input: string; base?: string }
  | {
      type: "room-start";
      cwd: string;
      harnessIds: string[];
      topic: string;
      rounds: number;
      mode?: "parallel" | "sequential";
      writeAllowed?: boolean;
      host?: { harnessId?: string; opening?: boolean; roundSummary?: boolean; finalSummary?: boolean };
      memberConfigs?: Record<string, Array<{ configId: string; value: string }>>;
    }
  | { type: "room-topic"; roomId: string; topic: string; rounds: number; mode?: "parallel" | "sequential" }
  | { type: "room-mode"; roomId: string; mode: "parallel" | "sequential" }
  | { type: "room-delete"; roomId: string; deleteSessions?: boolean }
  | { type: "room-run"; roomId: string }
  | { type: "room-stop"; roomId: string }
  | { type: "list" };

export type ServerMsg =
  | {
      type: "hello";
      harnesses: HarnessAvailability[];
      sessions: SessionInfo[];
      defaultCwd: string;
      rooms?: Room[];
      providers?: Array<{ id: string; label: string }>;
    }
  | { type: "session"; session: SessionInfo }
  | { type: "update"; sessionId: string; update: Record<string, unknown> }
  | { type: "turn_end"; sessionId: string; stopReason: string }
  | { type: "permission"; sessionId: string; requestId: string; title: string; options: PermissionOption[] }
  | { type: "transcript"; sessionId: string; entries: TranscriptEntry[] }
  | { type: "rooms"; rooms: Room[] }
  | { type: "room"; room: Room }
  | { type: "dirs" } & DirListing
  | { type: "history"; providers: Array<{ id: string; label: string }>; summaries?: Array<{ label: string; imported: number; updated: number; skipped: number }> }
  | { type: "log"; sessionId: string; line: string }
  | { type: "deleted"; sessionId: string }
  | { type: "error"; sessionId?: string; message: string };

File: app/lib/core/protocol.dart
/// WS 协议类型（与服务端 server/types.ts 对齐的移动端子集）。
/// APP 只做手机场景核心：会话列表 + 聊天 + 审批；编排/管理留给网页。
library;

// ---------- 服务端 → 客户端 ----------

class ConfigOption {
  final String id;
  final String? name;
  final String? currentValue;
  final List<({String value, String? name})> options;
  ConfigOption({required this.id, this.name, this.currentValue, this.options = const []});

  factory ConfigOption.fromJson(Map<String, dynamic> j) => ConfigOption(
        id: j['id'] as String,
        name: j['name'] as String?,
        currentValue: j['currentValue'] as String?,
        options: ((j['options'] as List<dynamic>?) ?? [])
            .map((o) => (value: o['value'] as String, name: o['name'] as String?))
            .toList(),
      );
}

class SessionInfo {
  final String id;
  final String harnessId;
  final String harnessLabel;
  final String cwd;
  final String status; // starting/ready/awaiting/saved/error/stopped
  final bool live;
  final bool resumable;
  final bool? inTurn;
  final bool? starred;
  final String? title;
  final String? autoApprove; // off/readonly/all
  final String? roomId;
  final String? error;
  final String lastActiveAt;
  final List<ConfigOption> configOptions;
  final PendingPermission? pendingPermission;

  SessionInfo({
    required this.id,
    required this.harnessId,
    required this.harnessLabel,
    required this.cwd,
    required this.status,
    required this.live,
    required this.resumable,
    this.inTurn,
    this.starred,
    this.title,
    this.autoApprove,
    this.roomId,
    this.error,
    required this.lastActiveAt,
    this.configOptions = const [],
    this.pendingPermission,
  });

  factory SessionInfo.fromJson(Map<String, dynamic> j) => SessionInfo(
        id: j['id'] as String,
        harnessId: j['harnessId'] as String,
        harnessLabel: j['harnessLabel'] as String? ?? '',
        cwd: j['cwd'] as String? ?? '',
        status: j['status'] as String? ?? 'saved',
        live: j['live'] as bool? ?? false,
        resumable: j['resumable'] as bool? ?? false,
        inTurn: j['inTurn'] as bool?,
        starred: j['starred'] as bool?,
        title: j['title'] as String?,
        autoApprove: j['autoApprove'] as String?,
        roomId: j['roomId'] as String?,
        error: j['error'] as String?,
        lastActiveAt: j['lastActiveAt'] as String? ?? '',
        configOptions: ((j['configOptions'] as List<dynamic>?) ?? [])
            .whereType<Map<String, dynamic>>()
            .map(ConfigOption.fromJson)
            .toList(),
        pendingPermission: j['pendingPermission'] == null
            ? null
            : PendingPermission.fromJson(j['pendingPermission'] as Map<String, dynamic>),
      );

  SessionInfo copyWith({String? status, bool? live, bool? inTurn, String? autoApprove, bool? starred, PendingPermission? pendingPermission, String? error, List<ConfigOption>? configOptions}) =>
      SessionInfo(
        id: id, harnessId: harnessId, harnessLabel: harnessLabel, cwd: cwd,
        status: status ?? this.status, live: live ?? this.live, resumable: resumable,
        inTurn: inTurn ?? this.inTurn, starred: starred ?? this.starred,
        title: title, autoApprove: autoApprove ?? this.autoApprove, roomId: roomId,
        error: error ?? this.error, lastActiveAt: lastActiveAt,
        configOptions: configOptions ?? this.configOptions,
        pendingPermission: pendingPermission ?? this.pendingPermission,
      );
}

class PendingPermission {
  final String requestId;
  final String title;
  final List<({String optionId, String name})> options;
  PendingPermission({required this.requestId, required this.title, required this.options});

  factory PendingPermission.fromJson(Map<String, dynamic> j) => PendingPermission(
        requestId: j['requestId'] as String,
        title: j['title'] as String? ?? '',
        options: ((j['options'] as List<dynamic>?) ?? [])
            .map((o) => (optionId: o['optionId'] as String, name: o['name'] as String))
            .toList(),
      );
}

/// 可用的 harness（hello 下发，用于新建会话时选择）
class HarnessInfo {
  final String id;
  final String label;
  final bool available;
  final String? state; // probed-ok/installed/needs-download/missing/blocked…
  final String? note;
  final bool experimental;
  final List<String> extraAgents;
  final List<ConfigOption> configs;
  HarnessInfo({
    required this.id,
    required this.label,
    required this.available,
    this.state,
    this.note,
    this.experimental = false,
    this.extraAgents = const [],
    this.configs = const [],
  });

  /// 探活通过（可用且真跑通过一次对话）
  bool get probedOk => available && state == 'probed-ok';

  factory HarnessInfo.fromJson(Map<String, dynamic> j) => HarnessInfo(
        id: j['id'] as String,
        label: j['label'] as String? ?? j['id'] as String,
        available: j['available'] as bool? ?? false,
        state: j['state'] as String?,
        note: j['note'] as String?,
        experimental: j['experimental'] as bool? ?? false,
        extraAgents: ((j['extraAgents'] as List<dynamic>?) ?? []).whereType<String>().toList(),
        configs: ((j['configs'] as List<dynamic>?) ?? [])
            .whereType<Map<String, dynamic>>()
            .map(ConfigOption.fromJson)
            .toList(),
      );
}

/// 台账条目（transcript 与实时 update 共用一套渲染）
class Entry {
  final String kind; // user/assistant/thought/tool/permission/error/log
  final String? text;
  final String? title; // tool/permission 用
  final String? status; // tool 用
  final String? toolCallId; // tool 用（更新时按 id 原地替换，避免重复行）
  final String? message; // error 用
  final String? answered; // permission 已答复
  final String? requestId; // 待审批
  final List<({String optionId, String name})>? options;
  Entry({required this.kind, this.text, this.title, this.status, this.toolCallId, this.message, this.answered, this.requestId, this.options});
}

// ---------- 客户端 → 服务端（构造原始 JSON map）----------

Map<String, dynamic> msgList() => {'type': 'list'};
Map<String, dynamic> msgPrompt(String sessionId, String text) => {'type': 'prompt', 'sessionId': sessionId, 'text': text};
Map<String, dynamic> msgTranscript(String sessionId) => {'type': 'transcript', 'sessionId': sessionId};
Map<String, dynamic> msgPermission(String sessionId, String requestId, String optionId) =>
    {'type': 'permission', 'sessionId': sessionId, 'requestId': requestId, 'optionId': optionId};
Map<String, dynamic> msgInterrupt(String sessionId) => {'type': 'interrupt', 'sessionId': sessionId};
Map<String, dynamic> msgResume(String sessionId) => {'type': 'resume', 'sessionId': sessionId};
Map<String, dynamic> msgClose(String sessionId) => {'type': 'close', 'sessionId': sessionId};
Map<String, dynamic> msgSetAutoApprove(String sessionId, String level) =>
    {'type': 'set-auto-approve', 'sessionId': sessionId, 'level': level};
Map<String, dynamic> msgConfig(String sessionId, String configId, String value) =>
    {'type': 'config', 'sessionId': sessionId, 'configId': configId, 'value': value};
Map<String, dynamic> msgStar(String sessionId, bool starred) =>
    {'type': 'star', 'sessionId': sessionId, 'starred': starred};
Map<String, dynamic> msgDelete(String sessionId) => {'type': 'delete', 'sessionId': sessionId};
Map<String, dynamic> msgCreate(String harnessId, {String? cwd, bool? isolate, Map<String, String>? vars}) => {
      'type': 'create',
      'harnessId': harnessId,
      if (cwd != null && cwd.isNotEmpty) 'cwd': cwd,
      if (isolate != null) 'isolate': isolate,
      if (vars != null && vars.isNotEmpty) 'vars': vars,
    };
